### PR TITLE
Sample/mission replay

### PR DESCRIPTION
**Pull request by ghost**
Refactored as MVC application with FXML

### CODE REVIEW

**Comment by ghost** [comment]
@john0005 Can you please review?

File: scene/mission-replay/src/main/java/com/esri/samples/scene/AnimationModel.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
+package com.esri.samples.scene;
```
Comment (john0005) [inline scene/mission-replay/src/main/java/com/esri/samples/scene/AnimationModel.java:1]: For Animation, Camera, and Plane model all that needs to be done is make sure that all methods have a description. :frowning:

File: scene/mission-replay/src/main/java/com/esri/samples/scene/AnimationModel.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    this.setKeyframe(keyframe);
+  }
+
+  public int getFrames() {
+    return frames.get();
+  }
+
+  /**
+   * Property tracking the number of frames in an animation.
+   * @return frames property
```
Comment (john0005) [inline scene/mission-replay/src/main/java/com/esri/samples/scene/AnimationModel.java:44]: Check that there is a space between description and tags

File: scene/mission-replay/src/main/java/com/esri/samples/scene/CameraModel.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+   * @param distance following distance for camera to view from
+   * @param angle following pitch angle for camera to view from. (0 - 90 degrees)
+   */
+  public CameraModel(boolean follow, double distance, double angle) {
+    this.follow = new SimpleBooleanProperty(follow);
+    this.distance = new SimpleDoubleProperty(distance);
+    this.angle = new SimpleDoubleProperty(angle);
+  }
+
+  public boolean getFollow() {
```
Comment (john0005) [inline scene/mission-replay/src/main/java/com/esri/samples/scene/CameraModel.java]: I would name this isFollowing

File: scene/mission-replay/src/main/resources/fxml/MissionView.fxml [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            <MapView fx:id="mapView">
+            </MapView>
+            <VBox StackPane.alignment="TOP_LEFT">
+                <Button text="+" onAction="#zoomInMap" prefWidth="25">
+                </Button>
+                <Button text="-" onAction="#zoomOutMap" prefWidth="25">
+                </Button>
+            </VBox>
+        </StackPane>
+    </VBox>
```
Comment (john0005) [inline scene/mission-replay/src/main/resources/fxml/MissionView.fxml]: Could add

``` FXML
<?import java.net.URL?>
<stylesheets> <URL value="@../style.css"/> </stylesheets>
```

Then take out `scene.getStylesheets().add(getClass().getResource("/style.css").toExternalForm());` from MissionReplay

File: scene/mission-replay/src/main/resources/fxml/MissionView.fxml [hunk 5]
```diff
@@ -0,0 +1,106 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<?import javafx.scene.layout.*?>
+
+<?import com.esri.arcgisruntime.mapping.view.SceneView?>
```
Comment (john0005) [inline scene/mission-replay/src/main/resources/fxml/MissionView.fxml]: will need to group these import for presentation purpose.

File: scene/mission-replay/src/main/resources/fxml/MissionView.fxml [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    </fx:define>
+    <!--SDK SceneView-->
+    <SceneView fx:id="sceneView">
+    </SceneView>
+    <!--Mission Selector Pane-->
+    <VBox StackPane.alignment="TOP_LEFT" spacing="5" maxWidth="200" maxHeight="50" styleClass="pane">
+        <ComboBox fx:id="missionSelector" promptText="Select a Mission" onAction="#changeMission">
+            <items>
+                <FXCollections fx:factory="observableArrayList">
+                    <String fx:value="GrandCanyon.csv">
```
Comment (john0005) [inline scene/mission-replay/src/main/resources/fxml/MissionView.fxml]: For these values that just take up one line we could do `<String fx:value="GrandCanyon.csv"/>` and then delete line 37
This way we can free up a lot of space.

File: scene/mission-replay/src/main/java/com/esri/samples/scene/MissionController.java [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+
+    } catch (Exception e) {
+      // on any exception, print the stack trace
+      e.printStackTrace();
+    }
+  }
+
+  /**
+   * Creates a 3D graphic representing the plane in the scene. Adds the graphic to the scene view's graphics overlay.
+   * @throws URISyntaxException if model cannot be loaded
```
Comment (john0005) [inline scene/mission-replay/src/main/java/com/esri/samples/scene/MissionController.java:163]: Need space between here

**Comment by john0005** [comment]
Some things to fix for the sample.
- change color of 2D plane symbol to blue so it's not red on red
- decrease the total zoom amount, goes out to far
- (+/-) buttons don't work when the plane is being animated (This worked today, so not a problem)
- need to default to the Grand canyon location when application first starts

Also on Mac when I try to run it the SceneView is completely black, will look more into that tomorrow.

Will look at MissionController tomorrow as well, only thing i have left to review.
Looks way better btw.

File: scene/mission-replay/src/main/java/com/esri/samples/scene/MissionController.java [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+   */
+  public void initialize() {
+
+    try {
+      // create a scene
+      ArcGISScene scene = new ArcGISScene(Basemap.createImagery());
+      sceneView.setArcGISScene(scene);
+
+      // set initial camera viewpoint
+      camera = new Camera(28.4, 83.9, 10010.0, 10.0, 80.0, 300.0);
```
Comment (john0005) [inline scene/mission-replay/src/main/java/com/esri/samples/scene/MissionController.java]: Will need to set this later to the starting location of grand canyon.

File: scene/mission-replay/src/main/java/com/esri/samples/scene/MissionController.java [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+    plane3D = new Graphic(new Point(0, 0, 0, SpatialReferences.getWgs84()), plane3DSymbol);
+    sceneOverlay.getGraphics().add(plane3D);
+  }
+
+  /**
+   * Creates a 2D graphic representing the plane on the mini map. Adds the graphic to the map view's graphics overlay.
+   */
+  private void create2DPlane() {
+    // create a red (0xFFFF0000) triangle symbol to represent the plane on the mini map
+    SimpleMarkerSymbol plane2DSymbol = new SimpleMarkerSymbol(SimpleMarkerSymbol.Style.TRIANGLE, 0xFFFF0000, 10);
```
Comment (john0005) [inline scene/mission-replay/src/main/java/com/esri/samples/scene/MissionController.java]: would change this to a blue

File: scene/mission-replay/src/main/java/com/esri/samples/scene/MissionController.java [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+      sceneView.setViewpointCamera(camera);
+
+      // add elevation data
+      Surface surface = new Surface();
+      surface.getElevationSources().add(new ArcGISTiledElevationSource(ELEVATION_IMAGE_SERVICE));
+      scene.setBaseSurface(surface);
+
+      // create a graphics overlay for the scene
+      sceneOverlay = new GraphicsOverlay();
+      sceneOverlay.setRenderer(new SimpleRenderer());
```
Comment (john0005) [inline scene/mission-replay/src/main/java/com/esri/samples/scene/MissionController.java]: don't believe this line is needed

**Comment by john0005** [comment]
Tried to find out what the problem with the black screen on Mac was today and through the process it stop doing it. I have heard of an issue were the SceneView will black screen sometimes when starting and that may have been the case. Will test it on linux to see how it runs.

Only on Mac the display for plane settings (lower right corner) is a little small and the words get cut off.
screen shot 2016-06-29 at 11 02 59 am

Also you can see that on Mac the planes heading, pitch, and roll are off not for sure if there is an issue for that or not.

**Comment by ghost** [comment]
@john0005 Second review please?

**Comment by ghost** [comment]
@john0005 One thing I didn't change is the max/min zoom on the mini map. I don't think this is a big issue because users have control here. I did notice though that if I set max and min scales on the map, I can still bypass them by setting the scale programatically. Is this a bug?

**Comment by john0005** [comment]
Everything looks good. About the scale thing thats a good question. I would bring up that same question here, https://devtopia.esri.com/runtime/runtime-mapping-api-design/issues, and let them figure it out. Seems like a bug to me.

File: scene/mission-replay/src/main/java/com/esri/samples/scene/AnimationModel.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+   * Constructs the animation model with the specified keyframe.
+   *
+   * @param keyframe starting animation frame
+   */
+  public AnimationModel(int keyframe) {
+    this();
+    this.setKeyframe(keyframe);
+  }
+
+  public int getFrames() {
```
Comment (vGandhi) [inline scene/mission-replay/src/main/java/com/esri/samples/scene/AnimationModel.java:37]: Some javadoc would be good.
~~Should it be getFrame() ? maybe getCurrentFrame() ?~~

File: scene/mission-replay/src/main/java/com/esri/samples/scene/CameraModel.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    this.following = new SimpleBooleanProperty();
+    this.distance = new SimpleDoubleProperty();
+    this.angle = new SimpleDoubleProperty();
+  }
+
+  /**
+   * Constructs the camera model with the specified property values.
+   *
+   * @param following if the camera should lock onto and following the target
+   * @param distance  followinging distance for camera to view from
```
Comment (vGandhi) [inline scene/mission-replay/src/main/java/com/esri/samples/scene/CameraModel.java]: typo

File: scene/mission-replay/src/main/java/com/esri/samples/scene/CameraModel.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+   * @param angle     followinging pitch angle for camera to view from. (0 - 90 degrees)
```
Comment (vGandhi) [inline scene/mission-replay/src/main/java/com/esri/samples/scene/CameraModel.java]: typo

File: scene/mission-replay/src/main/java/com/esri/samples/scene/MissionController.java [hunk 14]
```diff
@@ -0,0 +1,335 @@
...
+
+import com.esri.arcgisruntime.geometry.Point;
+import com.esri.arcgisruntime.geometry.PointCollection;
+import com.esri.arcgisruntime.geometry.Polyline;
+import com.esri.arcgisruntime.geometry.SpatialReferences;
+import com.esri.arcgisruntime.mapping.*;
+import com.esri.arcgisruntime.mapping.view.*;
+import com.esri.arcgisruntime.symbology.*;
+
+public class MissionController {
```
Comment (vGandhi) [inline scene/mission-replay/src/main/java/com/esri/samples/scene/MissionController.java:40]: Some javadoc stating that this is instantiated through FXML would be helpful for new JavafFx devs.

File: scene/mission-replay/src/main/java/com/esri/samples/scene/MissionController.java [hunk 15]
```diff
@@ -0,0 +1,335 @@
...
+  @FXML private ComboBox<String> missionSelector;
+  @FXML private Slider progressSlider;
+  @FXML private ToggleButton playButton;
+  @FXML private ToggleButton followButton;
+  @FXML private Slider zoomSlider;
+  @FXML private Slider angleSlider;
+  @FXML private Slider speedSlider;
+
+  private Camera camera;
+  private Timeline timer;
```
Comment (vGandhi) [inline scene/mission-replay/src/main/java/com/esri/samples/scene/MissionController.java]: Could we call it animationTimer?

File: scene/mission-replay/src/main/resources/fxml/MissionView.fxml [hunk 16]
```diff
@@ -0,0 +1,79 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (vGandhi) [inline scene/mission-replay/src/main/resources/fxml/MissionView.fxml]: Based on several examples from Oracle, I would name this file in lower-case, with underscore separator -
"mission_view.fxml", or just "mission.fxml". Android follows similar naming convention.

**Comment by vGandhi** [comment]
@tyle8552 few minor comments. Looks good overall.

File: scene/mission-replay/src/main/java/com/esri/samples/scene/MissionController.java (hunk 15 above)
Comment (ghost) [inline scene/mission-replay/src/main/java/com/esri/samples/scene/MissionController.java]: @vGandhi Or just `animation`?

Comment (vGandhi) [inline scene/mission-replay/src/main/java/com/esri/samples/scene/MissionController.java]: @tyle8552 that's good too.

**Comment by ghost** [comment]
Starting a new PR to address conflicts: https://github.com/Esri/arcgis-runtime-samples-java/pull/55

**Pull request closed without merging**